# Week 4

2023: Week 4 - New Customers
January 25, 2023
Challenge by: Jenny Martin

It's the final week of beginner month already! Hopefully you've picked up a lot of new skills and have enjoyed the challenges. We'll add another new technique and build on your existing skills with this challenge.

Data Source Bank acquires new customers every month. They are stored in separate tabs of an Excel workbook so it's "easy" to see which customers joined in which month. However, it's not so easy to do any comparisons between months. Therefore, we'd like to consolidate all the months into one dataset.

There's an extra twist as well. The customer demographics are stored as rows rather than columns, which doesn't make for very easy reading. So we'd also like to restructure the data.

Inputs

![image.png](attachment:image.png)


### Requirements
1. Input the data
2. We want to stack the tables on top of one another, since they have the same fields in each sheet. We can do this one of 2 ways (help):
- Drag each table into the canvas and use a union step to stack them on top of one another
- Use a wildcard union in the input step of one of the tables
- Some of the fields aren't matching up as we'd expect, due to differences in spelling. Merge these fields together
- Make a Joining Date field based on the Joining Day, Table Names and the year 2023

### Output

![image.png](attachment:image.png)

# New Section

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import os

In [ ]:
import pandas as pd

In [ ]:
file = '/content/drive/MyDrive/Colab Notebooks/Prepping Data/Week 4/New Customers.xlsx'

In [ ]:
f = pd.ExcelFile(file)

# Create an empty list to hold individual dataframes
list_of_dfs = []

# Loop through each worksheet
for sheet in f.sheet_names:

  #Parse each worksheet as a dataframe
  df = f.parse(sheet)

  # Add a column to indicate which sheet the individual df is from
  df['Table Name'] =sheet

  # Append dataframe to list
  list_of_dfs.append(df)

In [ ]:
# Concat (append) all the individual dfs
df = pd.concat(list_of_dfs, ignore_index=True)

In [ ]:
df.head()

ID  Joining Day    Demographic      Value Table Name Demographiic  \
0  490910            3      Ethnicity      White    January          NaN   
1  490910            3  Date of Birth  5/23/1981    January          NaN   
2  490910            3   Account Type      Basic    January          NaN   
3  369221           18      Ethnicity      Black    January          NaN   
4  369221           18  Date of Birth   3/4/2019    January          NaN   

  Demagraphic  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [ ]:
df.tail()

ID  Joining Day    Demographic       Value Table Name Demographiic  \
2965  174699            2  Date of Birth   3/13/1989   December          NaN   
2966  174699            2   Account Type        Gold   December          NaN   
2967  514598           28      Ethnicity       Other   December          NaN   
2968  514598           28  Date of Birth  10/10/1971   December          NaN   
2969  514598           28   Account Type    Platinum   December          NaN   

     Demagraphic  
2965         NaN  
2966         NaN  
2967         NaN  
2968         NaN  
2969         NaN

In [ ]:
cols = ['Demagraphic', 'Demographiic', 'Demographic']

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2970 entries, 0 to 2969
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ID            2970 non-null   int64 
 1   Joining Day   2970 non-null   int64 
 2   Demographic   2436 non-null   object
 3   Value         2970 non-null   object
 4   Table Name    2970 non-null   object
 5   Demographiic  258 non-null    object
 6   Demagraphic   276 non-null    object
dtypes: int64(2), object(5)
memory usage: 162.5+ KB


In [ ]:
df.head()

ID  Joining Day    Demographic      Value Table Name Demographiic  \
0  490910            3      Ethnicity      White    January          NaN   
1  490910            3  Date of Birth  5/23/1981    January          NaN   
2  490910            3   Account Type      Basic    January          NaN   
3  369221           18      Ethnicity      Black    January          NaN   
4  369221           18  Date of Birth   3/4/2019    January          NaN   

  Demagraphic  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [ ]:
df['Demographic'].fillna(df['Demographiic'], inplace=True)
df['Demographic'].fillna(df['Demagraphic'], inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2970 entries, 0 to 2969
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ID            2970 non-null   int64 
 1   Joining Day   2970 non-null   int64 
 2   Demographic   2970 non-null   object
 3   Value         2970 non-null   object
 4   Table Name    2970 non-null   object
 5   Demographiic  258 non-null    object
 6   Demagraphic   276 non-null    object
dtypes: int64(2), object(5)
memory usage: 162.5+ KB


In [ ]:
df = df[['ID', 'Joining Day', 'Demographic', 'Value','Table Name']]

In [ ]:
df.head()

ID  Joining Day    Demographic      Value Table Name
0  490910            3      Ethnicity      White    January
1  490910            3  Date of Birth  5/23/1981    January
2  490910            3   Account Type      Basic    January
3  369221           18      Ethnicity      Black    January
4  369221           18  Date of Birth   3/4/2019    January

In [ ]:
df['Joining Date'] = df['Joining Day'].astype('str') + " " + df['Table Name'] + " 2023"

<ipython-input-58-1bd7b61344a0>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Joining Date'] = df['Joining Day'].astype('str') + " " + df['Table Name'] + " 2023"


In [ ]:
df['Joining Date'] = pd.to_datetime(df['Joining Date'])

<ipython-input-59-37c036a4bc00>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Joining Date'] = pd.to_datetime(df['Joining Date'])


In [ ]:
df.head()

ID  Joining Day    Demographic      Value Table Name Joining Date
0  490910            3      Ethnicity      White    January   2023-01-03
1  490910            3  Date of Birth  5/23/1981    January   2023-01-03
2  490910            3   Account Type      Basic    January   2023-01-03
3  369221           18      Ethnicity      Black    January   2023-01-18
4  369221           18  Date of Birth   3/4/2019    January   2023-01-18

In [ ]:
 df = df.loc[:, ['ID', 'Demographic', 'Value', 'Joining Date']]

In [ ]:
df.head()

ID    Demographic      Value Joining Date
0  490910      Ethnicity      White   2023-01-03
1  490910  Date of Birth  5/23/1981   2023-01-03
2  490910   Account Type      Basic   2023-01-03
3  369221      Ethnicity      Black   2023-01-18
4  369221  Date of Birth   3/4/2019   2023-01-18


3.  Now we want to reshape our data so we have a field for each demographic, for each new customer (help)
4. Make sure all the data types are correct for each field
Remove duplicates (help)
If a customer appears multiple times take their earliest joining date

In [ ]:
df_flat = df.pivot(index=['ID', 'Joining Date'], columns ='Demographic', values ='Value').reset_index()

In [ ]:
df_flat.head()

Demographic      ID Joining Date Account Type Date of Birth Ethnicity
0            100185   2023-05-20        Basic     7/29/1952     Asian
1            101515   2023-04-14         Gold     8/11/1974     Black
2            101744   2023-08-29        Basic     1/21/1945     Asian
3            102704   2023-01-23        Basic      3/9/2000     Black
4            103488   2023-08-28        Basic     9/26/1957     Other

In [ ]:
df_flat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 990 entries, 0 to 989
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   ID             990 non-null    int64         
 1   Joining Date   990 non-null    datetime64[ns]
 2   Account Type   990 non-null    object        
 3   Date of Birth  990 non-null    object        
 4   Ethnicity      990 non-null    object        
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 38.8+ KB


In [ ]:
# Convert 'Date of Birth' to date
df_flat['Date of Birth'] = pd.to_datetime(df_flat['Date of Birth'], format="%m/%d/%Y")

In [ ]:
df_flat.head()

Demographic      ID Joining Date Account Type Date of Birth Ethnicity
0            100185   2023-05-20        Basic    1952-07-29     Asian
1            101515   2023-04-14         Gold    1974-08-11     Black
2            101744   2023-08-29        Basic    1945-01-21     Asian
3            102704   2023-01-23        Basic    2000-03-09     Black
4            103488   2023-08-28        Basic    1957-09-26     Other

In [ ]:
df_flat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 990 entries, 0 to 989
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   ID             990 non-null    int64         
 1   Joining Date   990 non-null    datetime64[ns]
 2   Account Type   990 non-null    object        
 3   Date of Birth  990 non-null    datetime64[ns]
 4   Ethnicity      990 non-null    object        
dtypes: datetime64[ns](2), int64(1), object(2)
memory usage: 38.8+ KB


In [ ]:
# Remove duplicates (help)
# If a customer appears multiple times take their earliest joining date

df_flat.sort_values(['ID', 'Joining Date'], inplace=True)
df_flat.head()

Demographic      ID Joining Date Account Type Date of Birth Ethnicity
0            100185   2023-05-20        Basic    1952-07-29     Asian
1            101515   2023-04-14         Gold    1974-08-11     Black
2            101744   2023-08-29        Basic    1945-01-21     Asian
3            102704   2023-01-23        Basic    2000-03-09     Black
4            103488   2023-08-28        Basic    1957-09-26     Other

In [ ]:
# check for duplicates ID
df_flat['ID'].duplicated().any()


True

In [ ]:
df_flat[df_flat.duplicated(['ID'])]

Demographic      ID Joining Date Account Type Date of Birth Ethnicity
852          878212   2023-12-22        Basic    1957-03-01     White

In [ ]:
df_flat.drop_duplicates(subset =['ID'], keep='first', inplace=True)

In [ ]:
df_flat.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 989 entries, 0 to 989
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   ID             989 non-null    int64         
 1   Joining Date   989 non-null    datetime64[ns]
 2   Account Type   989 non-null    object        
 3   Date of Birth  989 non-null    datetime64[ns]
 4   Ethnicity      989 non-null    object        
dtypes: datetime64[ns](2), int64(1), object(2)
memory usage: 46.4+ KB


Output

![image.png]()

5 fields
- ID
- Joining Date
- Account Type
- Date of Birth
- fEthnicity
989 rows (990 including headers)